benches/main.rs

```diff
@@ -0,0 +1,272 @@
+#![cfg(feature = "benches")]
+#![feature(test)]
+
+extern crate test;
+
+use anyhow::{bail, Context as _, Result};
+use std::path::Path;
+use wasmtime::{Config, Engine, OptLevel, Store, Strategy};
+use wasmtime_wast::WastContext;
+use wast::{
+    parser::{self, ParseBuffer},
+    Wat,
+};
+
+macro_rules! mk_test {
+    ($(#[$attrs:meta])* $name:ident, $path:expr, $strategy:expr) => {
+        mod $name {
+            use wasmtime::Strategy;
+
+            #[bench]
+            $(#[$attrs])*
+            fn compile(b: &mut ::test::bench::Bencher) {
+                crate::bench_compile(b, $path, $strategy).unwrap();
+            }
+
+            #[bench]
+            $(#[$attrs])*
+            fn run(b: &mut ::test::bench::Bencher) {
+                crate::bench_run(b, $path, $strategy).unwrap();
+            }
+        }
+    };
+}
+
+include!(concat!(env!("OUT_DIR"), "/wast_testsuite_tests.rs"));
+
+fn bench_compile(b: &mut test::bench::Bencher, wast: &str, strategy: Strategy) -> Result<()> {
+    let path = Path::new(wast);
+
+    let simd = path.iter().any(|s| s == "simd");
+
+    let bulk_mem = path.iter().any(|s| s == "bulk-memory-operations");
+
+    // Some simd tests assume support for multiple tables, which are introduced
+    // by reference types.
+    let reftypes = simd || path.iter().any(|s| s == "reference-types");
+
+    let multi_val = path.iter().any(|s| s == "multi-value");
+
+    let mut cfg = Config::new();
+    cfg.wasm_simd(simd)
+        .wasm_bulk_memory(bulk_mem)
+        .wasm_reference_types(reftypes)
+        .wasm_multi_value(multi_val)
+        .strategy(strategy)?
+        .cranelift_debug_verifier(cfg!(debug_assertions));
+
+    // FIXME: https://github.com/bytecodealliance/wasmtime/issues/1186
+    if simd {
+        cfg.cranelift_opt_level(OptLevel::None);
+    }
+
+    let store = Store::new(&Engine::new(&cfg));
+    let mut wast_context = WastContext::new(store);
+    wast_context.register_spectest()?;
+
+    let adjust_wast = |mut err: wast::Error| {
+        err.set_path(path);
+        err.set_text(wast);
+        err
+    };
+
+    let file_contents = std::fs::read_to_string(path)
+        .with_context(|| format!("failed to read `{}`", path.display()))?;
+    let buf = ParseBuffer::new(&file_contents).map_err(adjust_wast)?;
+    let ast = parser::parse::<wast::Wast>(&buf).map_err(adjust_wast)?;
```

Comment:
Could this be deduplicated with the `wasmtime-wast` crate? Perhaps something where a function to handle each directive is provided and then we say "run this file", or something like that?